components/net/bluetooth_test.rs

```diff
@@ -0,0 +1,541 @@
+const GLUCOSE_DEVICE_NAME: &'static str = "Glucose Device";
+const HEART_RATE_DEVICE_NAME: &'static str = "Heart Rate Device";
+const UNICODE_DEVICE_NAME: &'static str = "❤❤❤❤❤❤❤❤❤";
+
+// Device addresses
+const CONNECTABLE_DEVICE_ADDRESS: &'static str = "00:00:00:00:00:05";
+const GLUCOSE_DEVICE_ADDRESS: &'static str = "00:00:00:00:00:02";
+const HEART_RATE_DEVICE_ADDRESS: &'static str = "00:00:00:00:00:04";
+const UNICODE_DEVICE_ADDRESS: &'static str = "00:00:00:00:00:03";
+
+// Service UUIDs
+const BLACKLIST_TEST_SERVICE_UUID: &'static str = "611c954a-263b-4f4a-aab6-01ddb953f985";
+const DEVICE_INFORMATION_UUID: &'static str = "0000180a-0000-1000-8000-00805f9b34fb";
+const GENERIC_ACCESS_SERVICE_UUID: &'static str = "00001800-0000-1000-8000-00805f9b34fb";
+const GLUCOSE_SERVICE_UUID: &'static str = "00001808-0000-1000-8000-00805f9b34fb";
+const HEART_RATE_SERVICE_UUID: &'static str = "0000180d-0000-1000-8000-00805f9b34fb";
+const HUMAN_INTERFACE_DEVICE_SERVICE_UUID: &'static str = "00001812-0000-1000-8000-00805f9b34fb";
+const TX_POWER_SERVICE_UUID: &'static str = "00001804-0000-1000-8000-00805f9b34fb";
+
+// Characteristic UUIDs
+const BLACKLIST_EXCLUDE_READS_CHARACTERISTIC_UUID: &'static str = "bad1c9a2-9a5b-4015-8b60-1579bbbf2135";
+const BODY_SENSOR_LOCATION_CHARACTERISTIC_UUID: &'static str = "00002a38-0000-1000-8000-00805f9b34fb";
+const DEVICE_NAME_CHARACTERISTIC_UUID: &'static str = "00002a00-0000-1000-8000-00805f9b34fb";
+const HEART_RATE_MEASUREMENT_CHARACTERISTIC_UUID: &'static str = "00002a37-0000-1000-8000-00805f9b34fb";
+const PERIPHERAL_PRIVACY_FLAG_CHARACTERISTIC_UUID: &'static str = "00002a02-0000-1000-8000-00805f9b34fb";
+const SERIAL_NUMBER_STRING_UUID: &'static str = "00002a25-0000-1000-8000-00805f9b34fb";
+
+// Descriptor UUIDs
+const BLACKLIST_EXCLUDE_READS_DESCRIPTOR_UUID: &'static str = "aaaaaaaa-aaaa-1181-0510-810819516110";
+const BLACKLIST_DESCRIPTOR_UUID: &'static str = "07711111-6104-0970-7011-1107105110aaa";
+const CHARACTERISTIC_USER_DESCRIPTION_UUID: &'static str = "00002901-0000-1000-8000-00805f9b34fb";
+const CLIENT_CHARACTERISTIC_CONFIGURATION_UUID: &'static str = "00002902-0000-1000-8000-00805f9b34fb";
+const NUMBER_OF_DIGITALS_UUID: &'static str = "00002909-0000-1000-8000-00805f9b34fb";
+
+pub fn generate_id() -> String {
+    let mut id = String::new();
+    let mut generated = false;
+    let mut rng = rand::thread_rng();
+    while !generated {
+        id = rng.gen::<u32>().to_string();
+        CACHED_IDS.with(|cache|
+            if !cache.borrow().contains(&id) {
+                cache.borrow_mut().insert(id.clone());
+                generated = true;
+            }
+        );
+    }
+    id
+}
+
+pub fn set_attribute_or_return_error(result: Result<(), Box<Error>>, sender: &IpcSender<BluetoothResult<()>>) {
+    match result {
+        Ok(_) => (),
+        Err(_) => return drop(sender.send(Err(BluetoothError::Type(FAILED_SET_ERROR.to_string())))),
+    }
```

Comment:
I tried this with `unwrap_or` and it compiles, but then all of our WPT tests fails.
But we can shorten this:
``` rust
    if result.is_err() {
        return drop(sender.send(Err(BluetoothError::Type(FAILED_SET_ERROR.to_string()))))
    }
```